Thread scope through JS1 Week 2 (#294)
CodeYourFuture/curriculum: content/js1/blocks/scope/index.md

+++
title = '🔭 Scope'
headless = true
time = 20
facilitation = false
emoji= '❌'
[objectives]
1='Define scope'
+++

The function `convertToPercentage` will only be useful if we can access the `percentage` string that it creates. Otherwise, we won't be able to use the result of `convertToPercentage` in other parts of our code. We can play computer to check that the `prercentage` variable is being created. Somebody suggests trying to log the `percentage` variable outside

```js
const decimalNumber = 0.5;

function convertToPercentage() {
  const percentage = `${decimalNumber * 100}%`;
}

convertToPercentage(0.5);
console.log(percentage);
```

However if we run the code above, we get an error:

```console
ReferenceError: percentage is not defined
```

> We get an error because of {{<tooltip title="scope">}}Scope is a region of our code where certain variables are accessible.{{</tooltip>}}. When `convertToPercentage` is declared we also define a **function scope** - the region of code enclosed inside the `{}` of `convertToPercentage`'s function body. Any variables declared inside the body of the function

Author: I'm finding it tricky to define - I've had another go but still not satisfied either

+++
title = '🔭 Scope'
headless = true
time = 20
facilitation = false
emoji= '❌'
[objectives]
1='Define scope'
2='Explain what happens when a variable is referenced out of scope'
+++

The function `convertToPercentage` will only be useful if we can access the `percentage` string that it creates. Otherwise, we won't be able to use the result of `convertToPercentage` in other parts of our code. We can try accessing the `percentage` variable outside the function body like this:

```js {linenos=table,hl_lines=["8"],linenostart=1}
const decimalNumber = 0.5;

function convertToPercentage() {
  const percentage = `${decimalNumber * 100}%`;
}

convertToPercentage(0.5);
console.log(percentage);
```

However if we run the code above, we get an error:

```console
ReferenceError: percentage is not defined
```

> We get an error because of {{<tooltip title="scope">}}Scope refers to the rules that determine where we can refer to variables.{{</tooltip>}}. When we define `convertToPercentage` we also define a **local scope** - the region of code enclosed inside `convertToPercentage`'s function body. This region is `convertToPercentage`'s **local scope**. This means any variables we declare inside `convertToPercentage`'s **local scope** can only be accessed within this region. If we attempt to reference a variable outside the scope where it was declared, then get a `ReferenceError`.

Reviewer: ```suggestion
We get an error because of {{<tooltip **title="scope">}}Scope** means where variables are and what you can access. {{</tooltip>}}. When we define `convertToPercentage` we also define a **local scope** - the region of code enclosed inside `convertToPercentage`'s function body. This region is `convertToPercentage`'s **local scope**. This means any variables we declare inside `convertToPercentage`'s **local scope** can only be accessed within this region. If we attempt to reference a variable outside the scope where it was declared, then get a `ReferenceError`.
```
